### ISSUE TITLE
NA handling issue in xBal inferentials when there are strata and clusters

### ISSUE THREAD

**Issue by benthestatistician** [markmfredrickson/RItools#47]
To address (separately for each Covariate column):  
- [x]  Strata with no non-NA values shouldn't contribute anything
- [x]  ~~Strata without non-NA values in either Tx or Ctl shouldn't contribute either~~
- [x] Once descriptives are done, impute missings to stratum means, or to 0 if there's nothing in the stratum. This can be achieved by aligning on stratum means and then imputing all missings to 0. (As a separate matter, I think post alignment transforms will make most sense to users if they're applied prior to aggregation. So they might be applied immediately after these imputations.)
- [x]  ~~Create a path for users to run their own NA imputation routines on Design objects,  then pass along to `xBalance`~~

In brief, the original problem was that the following expectation recorded in Design.R's inline comments may not be met, in particular when one has a stratum containing clusters of unequal sizes and a variable that's all NAs:

```
 a <- slm.fit.csr.fixed(S, Covs)$residuals
 b <- slm.fit.csr.fixed(S, NotMiss)$fitted
 #this is needed for the case when a strata has all missing in either the treated or control group
 #setting this to the value closet to zero will give us a zero in the appropriate place in tmat, so all is good
b[b == 0] <- .Machine$double.eps
tmat <- a / b
```

NA's will have been replaced with imputation values prior to aggregation over clusters.  When clusters have different sizes, the result will be that aggregated values corresponding to the NAs differ, and the residuals recorded in `a` are not all zero, as the comment seems to expect. 

Example:

``` r
dat <- data.frame(strat=rep(letters[1:2], c(3,2)),
                  clus=factor(c(1,1,2:4)),
                  z=c(TRUE, rep(c(TRUE, FALSE), 2)),
                  x=rep(c(NA, 1), c(3,2))
                  )
design <- RItools:::makeDesign(z~x+strata(strat)+cluster(clus)-1, dat)
design <- RItools:::aggregateDesign(design)
Covs <- design@Covariates
NotMiss <- design@NotMissing
S <- SparseMMFromFactor(design@StrataFrame[,"strat"])
a <- RItools:::slm.fit.csr.fixed(S, Covs)$residuals
b <- RItools:::slm.fit.csr.fixed(S, NotMiss)$fitted
a[,1, drop=F] ; b[,1, drop=F]
a[,1, drop=F] / b[,1, drop=F]
```

**Comment by benthestatistician** [comment]
Commit 68cf3ce addresses one piece of the puzzle, the one pertaining to covariate/strata combinations for which all clusters are missing a value. It does this by changing the way covariates are aggregated across clusters: with this commit, when figuring its `Covariates` slot `aggregateDesign` only adds together non-NA values of a variable to get a cluster total. 

In consequence, for a stratum in which all observations of a covariate were NA, the aggregated Design object has 0's in the corresponding rows and columns of its Covariates slot.  In these situations, then, the expectation stated in the inline comment quoted in the issue statement will be met. (The data analysis example prompting this issue was one of those situations.)  

Remaining to be resolved:

1. I'm not certain this is the best handling of NAs for clusters with a combination of NA and non-NA values.  Perhaps instead such clusters should be represented by the mean of non-missing observations but the total number of observations, missing and non-missing.  I.e., aggregate across the cluster using the cluster mean of non-NA observations as an imputation value for missing observations.
2. After aggregation of the data (ie in `alignDesignByStrata`), how to handle covariate/stratum combinations that have a mixture of NA and non-NA cluster totals? Intuitively we'd want to drop the NA cluster totals and proceed if possible, ie if the stratum has non-NA clusters in both treatment and control groups, but otherwise drop the stratum. Since that would complicate combining the univariate imbalance measures into a chi-square value, however, I'd prefer to emulate this behavior by replacing the NA cluster totals with placeholder values of some kind.  A natural placeholder being the stratum mean of non-NA cluster totals.

**Comment by benthestatistician** [comment]
Hey @jwbowers and @markmfredrickson, how strong are your feelings about

1. having the `xBalance` default imputation be of medians rather than means
2. user-provided imputation schemes operating on the individual variable only, as opposed to on a matrix of covariates
3. allowing the user to override whatever imputation defaults we may come up with

(Indirectly related to this issue, as your answers may constrain the eventual resolution of this issue.)

**Comment by markmfredrickson** [comment]
I'm ok with all three. If users want to build models to impute data, they can do that before calling `xBalance`.

As a general rule, I'm all for imputation over complicated logic.

**Comment by jwbowers** [comment]
I don’t have strong feelings.

In general, allowing for openness in the API seems useful — mainly to us in the future, but perhaps also to others.

**Comment by benthestatistician** [comment]
Thanks guys!  @markmfredrickson, I agree mostly but not entirely about the possibility of users imputing data before calling `xBalance`.  What the users can't do, at least not without unwanted effort, is to do imputation on the numeric matrix representation of a factor variable.  That in turn can be a useful thing to do if your purpose is to impute in such a way that the adjusted mean differences you wind up with are the same as if you had done casewise deletion. The What Works Clearinghouse guidelines permit casewise deletion, although they specifically forbid the use of imputation for demonstrations of baseline equivalence.

This is part of the reason I was considering the option of mean imputation in all cases -- for other sorts of imputation, it seemed to me we could just rely on the user to do them.  In light of @jwbowers's  preference for an open API, however, I'll first explore options for giving the user some control here.

**Comment by markmfredrickson** [comment]
Maybe we could make use of the `fill.NAs` object work from `optmatch`. Eg.
calls like: `xBalance(fill.NAs(y ~ x + strata(z), ... stuff for imputation
..., data = mydata))`. What was nice about that is we created a subclass of
formula that could be hooked into later. As I recall, I made one proof of
concept commit, but never really followed up on the idea. For xBalance, we
might first strip the fill.NAs object of any `strata` and `cluster` calls,
and then let fill.NAs do it's thing.

-M

**Comment by benthestatistician** [comment]
I'm tempted to address these NA handling issues as part of a program to add support for users to provide sample inclusion weights along with the data, essentially using `model.frame`'s weights argument.  The outlines of that program are on view now in the clusters branch, in particular with  commitx 80404ff and 59ea96a.

**Comment by benthestatistician** [comment]
I think I was incorrect to say that "Strata without non-NA values in either Tx or Ctl shouldn't contribute".   (As of this writing, that's one of the checkboxes in this issue.)  Assume that NA-ness is independent of treatment status, and suppose a stratum in which there don't happen to be non-NA values in, say, the Tx group, although there are non-NA values. Seems more correct not to deny that stratum a contribution but instead to just score the treatment mean for that stratum as a 0. Denying a stratum a contribution to the total shouldn't be decided on the basis of a criterion like "no NAs in the treatment group" that might have worked out differently under a different treatment assignment.

**Comment by benthestatistician** [comment]
On code review, everything that still needed to be done in this issue was done, except for a tidbit addressed in a21312f6 and the parenthetical about applying post alignment transformations at the right time (which is now #64).  Closing.

**Comment by benthestatistician** [comment]
In order to address some other issues we're having, I want to adjust how this NA handling is done.  Briefly,

1.  For "inferentials" purposes, NAs should be imputed to stratum means. 
2. Then we need an unweighted alignment of cluster totals within strata. (As opposed to centering of cluster means around weighted stratum mean of cluster means.)

I think this can be done inside of `alignDesignsByStrata()`, without significant modification of inputs or outputs to that function.
